user_tasks/signals.py

```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+"""Celery signal handlers."""
+
+from __future__ import absolute_import, unicode_literals
+
+import logging
+from uuid import uuid4
+
+from django.contrib.auth import get_user_model
+from django.utils.module_loading import import_string
+
+from celery.signals import before_task_publish, task_failure, task_prerun, task_retry, task_success
+
+from .exceptions import TaskCanceledException
+from .models import UserTaskStatus
+from .tasks import UserTaskMixin
+
+LOGGER = logging.getLogger(__name__)
+
+
+@before_task_publish.connect
+def create_user_task(sender=None, body=None, **kwargs):  # pylint: disable=unused-argument
+    """
+    Create a :py:class:`UserTaskStatus` record for each :py:class:`UserTaskMixin`.
+
+    Also creates a :py:class:`UserTaskStatus` for each chain, chord, or group containing
+    the new :py:class:`UserTaskMixin`.
+    """
+    try:
+        task_class = import_string(sender)
+    except ImportError:
+        return
+    if issubclass(task_class.__class__, UserTaskMixin):
+        if 'user_id' in body['kwargs']:
+            user_id = body['kwargs']['user_id']
+        elif body['args']:
+            user_id = body['args'][0]
+        else:
+            raise TypeError('Each invocation of a UserTask must include the user_id')
+        try:
+
+            get_user_model().objects.get(pk=user_id)
+        except (ValueError, get_user_model().DoesNotExist):
+            raise TypeError('Invalid user_id: {}'.format(user_id))
+        task_id = body['id']
+        # Is it part of a chain?
+        if body.get('callbacks', []):
+            return create_chain_entry(user_id, task_id, task_class, body['args'], body['kwargs'], body['callbacks'])
+        if body.get('chord', None):
+            return create_chord_entry(task_id, task_class, body, user_id)
+        arguments_dict = task_class.arguments_as_dict(*body['args'], **body['kwargs'])
```

Comment:
Nit: Maybe this code would be clearer like this?:
```
if body.get('callbacks', []):
    return _create_chain_entry()
elif body.get('chord', None):
    return _create_chord_entry()
else:
    return _create_task_entry()
```
and breaking out the rest of the method into `_create_task_entry()`?